olp-cpp-sdk-core/include/olp/core/thread/TaskContinuation.h

```diff
@@ -0,0 +1,397 @@
+
+  using Task = std::function<OutResult(void*)>;
+  using Callback = std::function<void(void*)>;
+  using AsyncTask = std::function<void(void*, Callback)>;
+  using FinalCallback = std::function<void(void*, bool)>;
+
+  ContinuationImpl() {}
+
+  /// Constructs a continuation using the given async task
+  ContinuationImpl(std::shared_ptr<TaskScheduler> group,
+                   std::shared_ptr<ExecutionContext> context,
+                   std::pair<AsyncTask, Task>&& async);
+
+  ContinuationImpl Then(std::pair<AsyncTask, Task>&& async);
+
+  void Run(FinalCallback callback) const;
+
+  std::shared_ptr<ExecutionContext> GetContext() const;
+
+ private:
+  struct Processor;
+
+  using SharedStep = std::shared_ptr<const std::pair<AsyncTask, Task>>;
+
+  std::shared_ptr<olp::thread::TaskScheduler> m_scheduler;
+  std::shared_ptr<ExecutionContext> m_execution_context;
+  std::deque<std::pair<AsyncTask, Task>> deque_;
+};
+
+}  // namespace internal
+
+/// Template specialization for Continuation of void type. It has the same
+/// interface as the generic version.
+template <>
+class Continuation<void> {
+  template <typename OtherResultType>
+  friend class Continuation;
+
+ public:
+  using Result = void;
+
+  template <typename Callable>
+  Continuation<internal::AsyncResultType<Callable>> Then(Callable callable) {
+    using NewResultType = internal::AsyncResultType<Callable>;
+    using Function = internal::TypeToFunctionInput<NewResultType>;
+    return Then(std::function<void(Function)>(std::move(callable)));
+  }
+
+  Continuation() = delete;
+
+  template <typename NewType>
+  Continuation<internal::RemoveRefAndConst<NewType>> Then(
+      std::function<void(std::function<void(NewType)>)> async) {
+    return m_cont.Then(toAsyncTask({}, std::move(async)));
+  }
+
+  Continuation Then(std::function<void(std::function<void()>)> async);
+
+  using ContinuationImpl = internal::ContinuationImpl;
+  using OutResult = ContinuationImpl::OutResult;
+  using Task = ContinuationImpl::Task;
+  using Callback = ContinuationImpl::Callback;
+
+ private:
+  template <typename NewType>
+  static std::pair<ContinuationImpl::AsyncTask, ContinuationImpl::Task>
+  toAsyncTask(std::shared_ptr<ExecutionContext> context,
+              std::function<void(std::shared_ptr<ExecutionContext>,
+                                 std::function<void(NewType)>)>&& async) {
+    using NewResultType = internal::RemoveRefAndConst<NewType>;
+    auto asyncMoC = std::move(async);
+    return {[asyncMoC, context](void*, Callback callback) {
+              auto callbackMoC = std::move(callback);
+              asyncMoC(context, [callbackMoC](NewResultType input) {
+                callbackMoC(static_cast<NewResultType*>(&input));
+              });
+            },
+            [](void* input) {
+              NewResultType& in = *static_cast<NewResultType*>(input);
+              auto result = std::make_unique<NewResultType>(std::move(in));
```

Comment:
Moving from a reference might not work. Not sure how reference collapsing rules handle this, but if I read this table correct https://www.ibm.com/docs/en/xl-c-and-cpp-aix/13.1.2?topic=operators-reference-collapsing-c11 then it should collapse to lvalue reference.

Reply:
done